test/compress/dead-code.js

```diff
@@ -68,10 +68,9 @@ dead_code_constant_boolean_should_warn_more: {
         side_effects : true,
     };
     input: {
-        while (!((foo && bar) || (x + "0"))) {
+        while (!(foo || (x + "0"))) {
             console.log("unreachable");
             var foo;
-            function bar() {}
```

Comment:
Interesting idea. While adding complexity to `lib/parse.js` on the `harmony` branch sounds worrying, we may end up needing to do so if ECMAScript breaks forward-compatibility, i.e. code valid in the older version is no longer accepted in a latter version.

IMHO `parse()` serves a slightly different purpose than `acorn`, in that we don't need it to throw stringent syntax errors rather than obeying GIGO.